storage/remote/queue_manager.go

```diff
@@ -800,14 +919,33 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan sample) {
 			s.droppedOnHardShutdown.Add(uint32(droppedSamples))
 			return
 
+		case <-s.reshard:
+			select {
+			case sample := <-queue:
+				// Capture the remaining (in case) sample from the queue.
+				pendingSamples[nPending].Labels = labelsToLabelsProto(sample.labels, pendingSamples[nPending].Labels)
+				pendingSamples[nPending].Samples[0].Timestamp = sample.t
+				pendingSamples[nPending].Samples[0].Value = sample.v
+			default:
+			}
+			close(queue)
```

Comment:
Why are we closing the queue here? Readers should never close the channels (it might also be worthwhile to explicitly make `queue` read only).